### PR TITLE
Add type hints in `_device_datacls.py` and `coders.py` [unitaryHACK] (#16)

### PR DESCRIPTION
**Pull request by TripleR47**
Type checked on `_device_datacls.py` and `coders.py`. 

Was getting `Skipping analyzing module_name: found module but no type hints or library stubs` error on line 19 of `_device_datacls.py`, wasn't able to correct it, used `#type: ignore` to ignore the same.

### CODE REVIEW

**Comment by LaurentAjdnik** [comment]
> Was getting `Skipping analyzing module_name: found module but no type hints or library stubs` error on line 19 of `_device_datacls.py`, wasn't able to correct it, used `#type: ignore` to ignore the same.

Hi @TripleR47! You can use the `--ignore-missing-imports` in the command line and remove the `#type: ignore`.

This option is planned to be set as such in CI by @HGSilveri (see #168).

**Comment by TripleR47** [comment]
> > Was getting `Skipping analyzing module_name: found module but no type hints or library stubs` error on line 19 of `_device_datacls.py`, wasn't able to correct it, used `#type: ignore` to ignore the same.
> 
> Hi @TripleR47! You can use the `--ignore-missing-imports` in the command line and remove the `#type: ignore`.
> 
> This option is planned to be set as such in CI by @HGSilveri (see #168).

Ya, I overlooked that, thanks for letting me know! :smiley:

File: pulser/devices/_device_datacls.py [hunk 1]
```diff
@@ -47,31 +47,31 @@ class Device:
...
     interaction_coeff: float = 5008713.
 
-    def __post_init__(self):
+    def __post_init__(self) -> None:
         # Hack to override the docstring of an instance
         self.__dict__["__doc__"] = self._specs(for_docs=True)
 
     @property
-    def channels(self):
+    def channels(self) -> Dict:
```
Comment (HGSilveri) [inline pulser/devices/_device_datacls.py:55]: You can be a bit more specific here:
```suggestion
    def channels(self) -> Dict[str, Channel]:
```

File: pulser/devices/_device_datacls.py [hunk 2]
```diff
@@ -13,12 +13,12 @@
 # limitations under the License.
```
Comment (HGSilveri) [inline pulser/devices/_device_datacls.py:14]: Add `from __future__ import annotations`

File: pulser/devices/_device_datacls.py [hunk 3]
```diff
@@ -172,6 +172,6 @@ def _specs(self, for_docs=False):
 
         return "\n".join(lines + ch_lines)
 
-    def _to_dict(self):
+    def _to_dict(self) -> Dict:
```
Comment (HGSilveri) [inline pulser/devices/_device_datacls.py:175]: Specify that dict is `Dict[str, Any]`

File: pulser/json/coders.py [hunk 4]
```diff
@@ -21,9 +21,11 @@
 from pulser.json.utils import obj_to_dict
 from pulser.parametrized import Variable
 
+from typing import Dict, Any
```
Comment (HGSilveri) [inline pulser/json/coders.py:24]: This should be next to the built-in libraries imports (first block)

File: pulser/json/coders.py [hunk 5]
```diff
@@ -21,9 +21,11 @@
...
+
 
 class PulserEncoder(JSONEncoder):
-    def default(self, o):
+    def default(self, o: Any) -> Dict:
```
Comment (HGSilveri) [inline pulser/json/coders.py:28]: This one is also `Dict[str, Any]`

File: pulser/json/coders.py [hunk 6]
```diff
@@ -37,12 +39,12 @@ def default(self, o):
 
 
 class PulserDecoder(JSONDecoder):
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
         # TODO: Check version compatibility (stored at the Sequence level)
-        self.vars = {}
+        self.vars: Dict = {}
```
Comment (HGSilveri) [inline pulser/json/coders.py:44]:
```suggestion
        self.vars: Dict[str, Variable] = {}
```

File: pulser/json/coders.py [hunk 7]
```diff
@@ -37,12 +39,12 @@ def default(self, o):
...
         super().__init__(object_hook=self.object_hook, *args, **kwargs)
 
-    def object_hook(self, obj):
+    def object_hook(self, obj: Dict) -> Dict:
```
Comment (HGSilveri) [inline pulser/json/coders.py:47]:
```suggestion
    def object_hook(self, obj: Dict[str, Any]) -> Any:
```

**Review by HGSilveri — CHANGES_REQUESTED** [review]
Same comment as before: sometimes, you can be more specific, especially in the types of the dictionaries keys and values.

**Comment by TripleR47** [comment]
Updated the suggested changes.

File: pulser/json/coders.py [hunk 8]
```diff
@@ -12,20 +12,20 @@
...
 
 from pulser.json.utils import obj_to_dict
 from pulser.parametrized import Variable
 
-from typing import Dict, Any
-
 
 class PulserEncoder(JSONEncoder):
-    def default(self, o: Any) -> Dict:
+    def default(self, o: Any) -> Union[Dict[str, Any], Any]:
```
Comment (HGSilveri) [inline pulser/json/coders.py:28]: Doesn't just `Dict[str, Any]` work? A union with `Any` is a bit redundant, is the same as being `Any`

**Review by HGSilveri — COMMENTED** [review]
Alright, this one seems done, there's just one question to address.

File: pulser/json/coders.py (hunk 8 above)
Comment (TripleR47) [reply pulser/json/coders.py:28]: Without `Any` it's showing `error: Returning Any from function declared to return "Dict[str, Any]"` in following lines of code - 
* https://github.com/pasqal-io/Pulser/blob/954fef70885a498df8d747c430b4c7d528e93615/pulser/json/coders.py#L28
* https://github.com/pasqal-io/Pulser/blob/954fef70885a498df8d747c430b4c7d528e93615/pulser/json/coders.py#L36

The return types are similar to a dictionary. I think casting will be a better option.

Comment (TripleR47) [reply pulser/json/coders.py]: Done

**Review by HGSilveri — APPROVED** [review]
LGTM, nice job!

**Pull request merged**